lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp

```diff
@@ -26,123 +28,144 @@ using namespace swift;
 
 STATISTIC(NumEliminatedInsts, "number of removed instructions");
 
-static bool optimizeGuaranteedArgument(SILArgument *Arg,
-                                       OwnershipChecker &Checker) {
-  bool MadeChange = false;
+namespace {
 
-  // Gather all copy_value users of Arg.
-  llvm::SmallVector<CopyValueInst *, 4> Worklist;
-  for (auto *Op : Arg->getUses()) {
-    if (auto *CVI = dyn_cast<CopyValueInst>(Op->getUser())) {
-      Worklist.push_back(CVI);
-    }
-  }
+struct SemanticARCOptVisitor
+    : SILInstructionVisitor<SemanticARCOptVisitor, bool> {
+  bool visitSILInstruction(SILInstruction *i) { return false; }
+  bool visitCopyValueInst(CopyValueInst *cvi);
+  bool visitBeginBorrowInst(BeginBorrowInst *bbi);
+};
 
-  // Then until we run out of copies...
-  while (!Worklist.empty()) {
-    auto *CVI = Worklist.pop_back_val();
-
-    // Quickly see if copy has only one use and that use is a destroy_value. In
-    // such a case, we can always eliminate both the copy and the destroy.
-    if (auto *Op = CVI->getSingleUse()) {
-      if (auto *DVI = dyn_cast<DestroyValueInst>(Op->getUser())) {
-        DVI->eraseFromParent();
-        CVI->eraseFromParent();
-        NumEliminatedInsts += 2;
+} // end anonymous namespace
+
+bool SemanticARCOptVisitor::visitBeginBorrowInst(BeginBorrowInst *bbi) {
+  auto kind = bbi->getOperand().getOwnershipKind();
+  SmallVector<EndBorrowInst *, 16> endBorrows;
+  for (auto *op : bbi->getUses()) {
+    auto *user = op->getUser();
+    switch (user->getKind()) {
+    case SILInstructionKind::EndBorrowInst:
+      endBorrows.push_back(cast<EndBorrowInst>(user));
+      break;
+    default:
+      // Make sure that this operand can accept our arguments kind.
+      auto map = op->getOwnershipKindMap();
+      if (map.canAcceptKind(kind))
```

Comment:
[design] I still don't understand what an OwnershipKindMap is for. I think it obscures the meaning of the code. Why not just:

`op->canAcceptOwnership(kind)`

Also, this check doesn't make sense to me for forwarding instructions, without teaching the optimization to look through all of those. If the forwarding instruction is not taking ownership before, we don't want to implicitly change it's meaning by changing the ownership of its incoming value.

This is what would make sense to me, instead of the map and "can accept" check.
```
// Is this a point-in-time use that cannot extend the lifetime if it's
// operand value?
if (op->isNonConsuming())
  continue;

return false
```